extension-impl/remoting-bolt/src/main/java/com/alipay/sofa/rpc/transport/bolt/ReuseBoltClientConnectionManager.java

```diff
@@ -0,0 +1,198 @@
+                LOGGER.warn("get connection failed in url," + url);
+            }
+            if (connection == null) {
+                return null;
+            }
+            // 保存唯一长连接
+            Connection oldConnection = urlConnectionMap.putIfAbsent(transportConfig, connection);
+            if (oldConnection != null) {
+                if (LOGGER.isWarnEnabled()) {
+                    LOGGER.warn("Multiple threads init ClientTransport with same key:" + url);
+                }
+                connection.close(); //如果同时有人插入，则使用第一个
+                connection = oldConnection;
+            } else {
+
+                // 增加计数器
+                AtomicInteger counter = connectionRefCounter.get(connection);
+                if (counter == null) {
+                    counter = new AtomicInteger(0);
+                    AtomicInteger oldCounter = connectionRefCounter.putIfAbsent(connection, counter);
+                    if (oldCounter != null) {
+                        counter = oldCounter;
+                    }
+                }
+                int currentCount = counter.incrementAndGet();
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug("Bolt client transport {} of {}, current ref count is: {}", url.toString(),
+                        NetUtils.channelToString(connection.getLocalAddress(), connection.getRemoteAddress()),
+                        currentCount);
+                }
+            }
+        }
+        return connection;
+    }
+
+    /**
+     * 关闭长连接
+     *
+     * @param rpcClient       bolt客户端
+     * @param transportConfig 传输层配置
+     * @param url             传输层地址
+     */
+    @Override
+    public void closeConnection(RpcClient rpcClient, ClientTransportConfig transportConfig, Url url) {
+        if (rpcClient == null || transportConfig == null || url == null) {
+            return;
+        }
+        // 先删除
+        Connection connection = urlConnectionMap.remove(transportConfig);
+        if (connection == null) {
+            return;
+        }
+        // 再判断是否需要关闭
+        boolean needDestroy;
+        AtomicInteger integer = connectionRefCounter.get(connection);
+        if (integer == null) {
+            needDestroy = true;
+        } else {
+            // 当前连接引用数
+            int currentCount = integer.decrementAndGet();
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Client transport {} of {} , current ref count is: {}", url.toString(),
+                    NetUtils.channelToString(connection.getLocalAddress(), connection.getRemoteAddress()),
+                    currentCount);
+            }
+            if (currentCount <= 0) {
+                // 此长连接无任何引用，可以销毁
+                connectionRefCounter.remove(connection);
+                needDestroy = true;
+            } else {
+                needDestroy = false;
+            }
+        }
+        if (needDestroy) {
+            rpcClient.closeConnection(url);
```

Comment:
This operate is dangerous, before `rpcClient.closeConnection(url)` invoke, may have new ref get this connection in concurrent scene.

Reply:
fixed